include/secp256k1_bulletproofs.h

```diff
@@ -0,0 +1,144 @@
+ *                   n: number of NUMS generators to produce. Should be 128 to allow for
+ *                      64-bit rangeproofs
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT secp256k1_bulletproofs_generators *secp256k1_bulletproofs_generators_create(
+    const secp256k1_context* ctx,
+    size_t n
+) SECP256K1_ARG_NONNULL(1);
+
+/** Allocates a list of generators from a static array
+ *  Returns a list of generators, or NULL if allocation or parsing failed.
+ *  Args:      ctx: pointer to a context object (cannot be NULL)
+ *  In:       data: data that came from `secp256k1_bulletproofs_generators_serialize`
+ *        data_len: the length of the `data` buffer
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT secp256k1_bulletproofs_generators* secp256k1_bulletproofs_generators_parse(
+    const secp256k1_context* ctx,
+    const unsigned char* data,
+    size_t data_len
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2);
+
+/** Serializes a list of generators to an array
+ *  Returns 1 on success, 0 if the provided array was not large enough
+ *  Args:        ctx: pointer to a context object (cannot be NULL)
+ *               gen: pointer to the generator set to be serialized (cannot be NULL)
+ *  In:         data: data that came from `secp256k1_bulletproofs_generators_serialize`
+ *  In/Out: data_len: the length of the `data` buffer. Should be initially set to at
+ *                    least 33 times the number of generators; will be set to 33 times
+ *                    the number of generators on successful return (cannot be NULL)
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_bulletproofs_generators_serialize(
+    const secp256k1_context* ctx,
+    secp256k1_bulletproofs_generators* gen,
+    unsigned char* data,
+    size_t *data_len
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2);
+
+/** Destroys a list of NUMS generators, freeing allocated memory
+ *  Args:   ctx: pointer to a context object (cannot be NULL)
+ *          gen: pointer to the generator set to be destroyed
+ */
+SECP256K1_API void secp256k1_bulletproofs_generators_destroy(
+    const secp256k1_context* ctx,
+    secp256k1_bulletproofs_generators* gen
+) SECP256K1_ARG_NONNULL(1);
+
+/** Returns the serialized size of an uncompressed proof of a given number of bits
+ *  Args:   ctx: pointer to a context object (cannot be NULL)
+ *  In:  n_bits: number of bits to prove (max 64, should usually be 64)
+ */
+SECP256K1_API size_t secp256k1_bulletproofs_rangeproof_uncompressed_proof_length(
+    const secp256k1_context* ctx,
+    size_t n_bits
+) SECP256K1_ARG_NONNULL(1);
+
+/** Produces an uncompressed rangeproof. Returns 1 on success, 0 on failure.
+ *  Args:   ctx: pointer to a context object (cannot be NULL)
+ *               gens: pointer to the generator set to use, which must have 2*n_bits generators (cannot be NULL)
+ *    asset_gen: pointer to the asset generator for the CT commitment (cannot be NULL)
```

Comment:
Is the concept of an "asset generator" still relevant in the context of bulletproofs?

If not, would it be safer to create one ourselves, instead of risking that the user passes in an insecure (known discrete log relation with others) generator?

Reply:
Yes, BPs don't change anything about how we manipulate assets.